src/compiler/checker.ts

```diff
@@ -20770,6 +20770,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const sourceType = i === restIndex ? getRestOrAnyTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
             const targetType = i === restIndex ? getRestOrAnyTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
             if (sourceType && targetType) {
+                if (sourceType === targetType) {
+                    continue;
+                }

```

Comment:
I'm not super fond of this as a fix. It could be treated as a potential perf optimization~ - although I really doubt it would show anything in the perf runs, it's just an early bailout from this loop. It just happens to fix this particular circularity.

Similar bailouts are used in a couple of places in the code (even at the top of the containing function - `compareSignaturesRelated`). 

Ideally, I'd just like to recognize this type as one without call signatures so it could be typechecked using regular rules. At first, I thought that this code could just be simplified for the scenarios with `strictVariance` - after all, if we already have strict variance on why do we need to special case anything here? 

It turns out though that despite strict variance the callback parameters are still special - even if they are bivariant methods (!). This is somewhat surprising and I just learned about it today, while working on this but I guess it makes sense. Covariant type checking for callback parameters was introduced [here](https://github.com/microsoft/TypeScript/pull/15104). You can also see what changed when I tried to ignore this logic [here](https://github.com/microsoft/TypeScript/commit/4777d6eed461ae29d5e554df08b71ea709552e37)

The problem arises on `main` because both `getNonNullableType` and `getSingleCallSignature` resolve properties of the type and both are called to learn if the parameter is a callback parameter.

In this specific case, I could also experiment with:
- avoiding transitive `resolveStructuredTypeMembers` in `getTypeFactsWorker` based on `callerOnlyNeeds`. if the caller only needs to learn if the type is nullable or not... we don't need to check if it's a function type to compute those facts
- avoiding `resolveStructuredTypeMembers` in `getSingleSignature` when `SignatureKind.Call` is requested and the type's symbol has a `SymbolFlags.Class`. If the type is a class then it has a construct signature (and thus `getSingleCallSignature` can't return `true` in such a case) so the code could bail out early.

Some extra consideration for unions/intersections could be put into the above too.

Reply:
cc @ahejlsberg - i was elaborating on this down here when u asked the question about what exact code has caused this ;p